### PR TITLE
Task 4: Integration with Database

### PR DESCRIPTION
**Pull request by siargei-stepanov**
FE PR: https://github.com/siargei-stepanov/shop-car-front/pull/3

Done:
- [x]  Created a DB in RDS service
- [x] Prepared SQL script to fill the DB with sample data (db.sql)
- [x] Integrated `getProductsList` and `getProductById` lambda functions with DB
- [x] Implemented `createProduct` lambda function in integration with DB
- [x] Return 400 status code if product data is invalid
- [x] All lambdas return error 500 status code on any error
- [x] All lambdas do console.log for each incoming requests and their arguments
- [x] Transactions

API endpoints:
* GetProductsList: GET: https://9kpocb3x23.execute-api.eu-west-1.amazonaws.com/products
* GetProductById: GET: https://9kpocb3x23.execute-api.eu-west-1.amazonaws.com/products/bd43a441-ea29-44ac-844b-35fd412b159c
* CreateProduct POST: https://9kpocb3x23.execute-api.eu-west-1.amazonaws.com/products
Sample product:
`{"manufacturer": "Alfa Romeo", "model":"Giulietta", "img":"https://www.alfaromeo.am/wp-content/uploads/2019/12/Alfa-Romeo_Giulia_720x404.jpg", "price":"26000", "count":"7"}`

FE:
https://d2tjeg0inghqta.cloudfront.net/

### CODE REVIEW

**Comment by ambroseus** [comment]
Hi @siargei-stepanov  we have 3 days till review deadline, so you still can implement transactions, for example
```ts
import { Pool, QueryConfig, QueryResult, PoolClient } from 'pg'

const pool = new Pool({
  host: process.env.DB_HOST,
  port: Number(process.env.DB_PORT),
  user: process.env.DB_USERNAME,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_DATABASE,
})

export async function dbTransaction<Result = any>(execFn: (client: PoolClient) => Promise<Result>) {
  const client = await pool.connect()
  let res: Result
  try {
    await client.query('BEGIN')
    try {
      res = await execFn(client)
      await client.query('COMMIT')
    } catch (err) {
      await client.query('ROLLBACK')
      throw err
    }
  } finally {
    client.release()
  }
  return res
}

export async function dbQuery<Row = any>(query: string | QueryConfig<any[]>, values?: any[]) {
  return await dbTransaction<QueryResult<Row>>((client) => client.query(query, values))
}
```

**Comment by ambroseus** [comment]
and there is a CORS error when I try to create product via FE
Screenshot from 2022-07-27 12-21-13

**Comment by siargei-stepanov** [comment]
> and there is a CORS error when I try to create product via FE Screenshot from 2022-07-27 12-21-13

I forgot to integrate the functionality of product creation on FE side. Will work on this. 
But the API itself is working using Postman

**Comment by ambroseus** [comment]
@siargei-stepanov  don't worry, you still have time to update your PRs to get more score points till task4 review deadline :)

**Comment by ambroseus** [comment]
@siargei-stepanov  about CORS - you can try this solution https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors-console.html

**Comment by siargei-stepanov** [comment]
> @siargei-stepanov about CORS - you can try this solution https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors-console.html

I've found how to fix the CORS issue properly, now it is correctly integrated into FE. The corresponding changes are already add to FE repo and link to the PR added to the description

**Comment by siargei-stepanov** [comment]
@ambroseus you can check the updated code with transactions and CORS enabled. Thank you for your comments!

File: product-service/handlers/createProduct.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    } catch (error) {
+        return {
+            statusCode: 400,
+            body: "Cannot parse request body",
+        };
+    }
+
+    return executeTransaction(async (client) => {
+        const {manufacturer, model, img, price, count} = product
+        const queryProduct = `insert into products (manufacturer, model, img, price) values
```
Comment (maksym0797) [inline product-service/handlers/createProduct.js:24]: [Parameterized query](https://node-postgres.com/features/queries#parameterized-query) should be used for protection from sql injections.

Comment (siargei-stepanov) [reply product-service/handlers/createProduct.js:24]: Fixed

**Review by ambroseus — APPROVED** [review]

File: product-service/serverless.yml [hunk 2]
```diff
@@ -45,8 +47,13 @@ provider:
...
 # you can define service wide environment variables here
-#  environment:
-#    variable1: value1
+  environment:
+    PG_HOST: task4-db.cqa5v273sikm.eu-west-1.rds.amazonaws.com
+    PG_PORT: 5432
+    PG_DB: task4
+    PG_USER: postgres
+    PG_PWD: 123
+
```
Comment (ambroseus) [inline product-service/serverless.yml:56]: please do not store secrets in config file in repo - use .env
Screenshot from 2022-07-31 12-49-23

**Review by ambroseus — APPROVED** [review]

File: product-service/serverless.yml (hunk 2 above)
Comment (ambroseus) [reply product-service/serverless.yml:56]: meanwhile you've got max score 9 from my side :)

**Pull request merged**